File: FractalScala/src/main/scala/de/gdietz/fun/fractal/scala/util/Quaternion.scala
package de.gdietz.fun.fractal.scala.util

import de.gdietz.fun.fractal.util.{ComplexNumberFormatter, Coordinate4D, Quaternion => JavaQuaternion}
import de.gdietz.util.NumberFormatHelper

import java.text.NumberFormat
import scala.language.implicitConversions

sealed trait OptQuaternion
  extends OptHigherNumber[OptQuaternion, Quaternion]
    with Product with Serializable {

  override def none: OptQuaternion = NoQuaternion
  override def zero: Quaternion = Quaternion.zero
  override def unit: Quaternion = Quaternion.one

  def toRealVector: RealVector4

}

object OptQuaternion {

  def apply(x: Double, y: Double = 0.0, z: Double = 0.0, w: Double = 0.0): OptQuaternion =
    Quaternion(x, y, z, w)

  val none: OptQuaternion = NoQuaternion


  implicit def doubleToOptQuaternion(x: Double): OptQuaternion =
    Quaternion(x)

  implicit def optRealToOptQuaternion(x: OptReal): OptQuaternion =
    x match {
      case Real(x) => Quaternion(x)
      case _ => NoQuaternion
    }

  implicit def optComplexToOptQuaternion(c: OptComplex): OptQuaternion =
    c match {
      case Complex(x, y) => Quaternion(x, y)
      case _ => NoQuaternion
    }

  implicit def coordinateToOptQuaternion(c: Coordinate4D): OptQuaternion =
    Quaternion(c.getX, c.getY, c.getZ, c.getW)

}


case class Quaternion(x: Double, y: Double = 0.0, z: Double = 0.0, w: Double = 0.0)
  extends OptQuaternion
    with SomeHigherNumber[OptQuaternion, Quaternion]
    with HigherHoloNumber[Quaternion] {

  override def isZero: Boolean = x == 0.0 && y == 0.0 && z == 0.0 && w == 0.0

  override def isUnit: Boolean = x == 1.0 && y == 0.0 && z == 0.0 && w == 0.0

  override def normSqr: Double = x * x + y * y + z * z + w * w

  def imagNormSqr: Double = y * y + z * z + w * w

  def imagNorm: Double = Math.sqrt(imagNormSqr)

  override def unary_- : Quaternion = Quaternion(-x, -y, -z, -w)

  override def +(r: Double): Quaternion = Quaternion(x + r, y, z, w)

  override def +(q: Quaternion): Quaternion = Quaternion(x + q.x, y + q.y, z + q.z, w + q.w)

  override def -(r: Double): Quaternion = Quaternion(x - r, y, z, w)

  override def -:(r: Double): Quaternion = Quaternion(r - x, -y, -z, -w)

  override def -(q: Quaternion): Quaternion = Quaternion(x - q.x, y - q.y, z - q.z, w - q.w)

  override def *(r: Double): Quaternion = Quaternion(r * x, r * y, r * z, r * w)

  override def *(q: Quaternion): Quaternion =
    Quaternion(x * q.x - y * q.y - z * q.z - w * q.w,
      x * q.y + y * q.x + z * q.w - w * q.z,
      x * q.z + z * q.x + w * q.y - y * q.w,
      x * q.w + w * q.x + y * q.z - z * q.y)

  def inverse: Quaternion = {
    val den: Double = normSqr
    Quaternion(x / den, -y / den, -z / den, -w / den)
  }

  override def /(r: Double): Quaternion = Quaternion(x / r, y / r, z / r, w / r)

  def conjugate: Quaternion = Quaternion(x, -y, -z, -w)

  override def sqr: Quaternion = {
    val hy: Double = x * y
    val hz: Double = x * z
    val hw: Double = x * w
    Quaternion(x * x - y * y - z * z - w * w, hy + hy, hz + hz, hw + hw)
  }

  override def cube: Quaternion = {
    val x2: Double = x * x
    val y2: Double = y * y
    val z2: Double = z * z
    val w2: Double = w * w
    val si: Double = y2 + z2 + w2
    Quaternion(x * (x2 - 3.0 * si), (3.0 * x2 - si) * y, (3.0 * x2 - si) * z, (3.0 * x2 - si) * w)
  }


  def iTimes: Quaternion =
    Quaternion(-y, x, -w, z)

  def timesI: Quaternion =
    Quaternion(-y, x, w, -z)

  def jTimes: Quaternion =
    Quaternion(-z, w, x ,-y)

  def timesJ: Quaternion =
    Quaternion(-z, -w, x, y)

  def kTimes: Quaternion =
    Quaternion(-w, -z, y, x)

  def timesK: Quaternion =
    Quaternion(-w, z, -y, x)


  /**
   * A power series `f` with real coefficients as a function on the complex plane can be extended to the quaternions.
   */
  def realPowerSeries(f: Complex => Complex): Quaternion = {
    val in = imagNorm
    val c = Complex(x, imagNorm)
    val fc = f(c)
    if (in == 0d)
      fc.toQuaternion   // f(c) should be real since c is real
    else {
      val q = fc.y / in
      Quaternion(fc.x, y * q, z * q, w * q)
    }
  }


  @inline override def exp: Quaternion = realPowerSeries(_.exp)
  @inline override def log: Quaternion = realPowerSeries(_.log)
  @inline override def log(branch: Int): Quaternion = realPowerSeries(_.log(branch))

  @inline override def sqrt: Quaternion = realPowerSeries(_.sqrt)

  @inline override def sin: Quaternion = realPowerSeries(_.sin)
  @inline override def cos: Quaternion = realPowerSeries(_.cos)
  @inline override def sinh: Quaternion = realPowerSeries(_.sinh)
  @inline override def cosh: Quaternion = realPowerSeries(_.cosh)
  @inline override def tan: Quaternion = realPowerSeries(_.tan)
  @inline override def tanh: Quaternion = realPowerSeries(_.tanh)


  override def toRealVector: RealVector4 =
    RealVector4(Real(x), Real(y), Real(z), Real(w))

  def toJavaQuaternion: JavaQuaternion = new JavaQuaternion(x, y, z, w)


  def toString(nf: NumberFormat,
               parentheses: Boolean = false,
               negativeParentheses: Boolean = false,
               sign: Boolean = false): String = {
    val cnf = new ComplexNumberFormatter(nf, parentheses, negativeParentheses, sign)
    cnf.addNumberString(x)
    cnf.addNumberString(y, "i")
    cnf.addNumberString(z, "j")
    cnf.addNumberString(w, "k")
    cnf.toString
  }

  override def toString: String = toString(NumberFormatHelper.getDefaultNumberFormat)

}


object Quaternion {

  val zero: Quaternion = Quaternion(0.0)
  val one: Quaternion = Quaternion(1.0)
  val i: Quaternion = Quaternion(0.0, 1.0)
  val j: Quaternion = Quaternion(0.0, 0.0, 1.0)
  val k: Quaternion = Quaternion(0.0, 0.0, 0.0, 1.0)
  val minusOne: Quaternion = Quaternion(-1.0)
  val minusI: Quaternion = Quaternion(0.0, -1.0)
  val minusJ: Quaternion = Quaternion(0.0, 0.0, -1.0)
  val minusK: Quaternion = Quaternion(0.0, 0.0, 0.0, -1.0)


  implicit def doubleToQuaternion(x: Double): Quaternion =
    Quaternion(x)

  implicit def realToQuaternion(x: Real): Quaternion =
    Quaternion(x.x)

  implicit def complexToQuaternion(c: Complex): Quaternion =
    Quaternion(c.x, c.y)

  implicit def coordinateToQuaternion(c: Coordinate4D): Quaternion =
    Quaternion(c.getX, c.getY, c.getZ, c.getW)

}

case object NoQuaternion
  extends OptQuaternion
    with NoHigherNumber[OptQuaternion, Quaternion] {

  override def toRealVector: RealVector4 =
    RealVector4(OptReal.none, OptReal.none, OptReal.none, OptReal.none)

  override def toString: String = "na"

}

File: Fractal/src/main/java/de/gdietz/fun/fractal/util/Quaternion.java
package de.gdietz.fun.fractal.util;

import de.gdietz.util.NumberFormatHelper;

import java.text.NumberFormat;

public class Quaternion extends Coordinate4D implements NormedNumber<Quaternion> {

    public static final Quaternion ZERO = new Quaternion();
    public static final Quaternion ONE = new Quaternion(1.0);
    public static final Quaternion I = new Quaternion(0.0, 1.0);
    public static final Quaternion J = new Quaternion(0.0, 0.0, 1.0, 0.0);
    public static final Quaternion K = new Quaternion(0.0, 0.0, 0.0, 1.0);
    public static final Quaternion MINUS_ONE = new Quaternion(-1.0);

    public static final Quaternion NaN = new Quaternion(Double.NaN, Double.NaN, Double.NaN, Double.NaN);

    public Quaternion() {
    }

    public Quaternion(double x) {
        super(x, 0.0, 0.0, 0.0);
    }

    public Quaternion(double x, double y) {
        super(x, y, 0.0, 0.0);
    }

    public Quaternion(double x, double y, double yj, double yk) {
        super(x, y, yj, yk);
    }

    public Quaternion(Coordinate c) {
        this(c.x, c.y);
    }

    public Quaternion(Quaternion q) {
        super(q);
    }

    public double getReal() {return x; }
    public double getImagI() {return y; }
    public double getImagJ() {return z; }
    public double getImagK() {return w; }

    public double imagNormSqr() {
		return y * y + z * z + w * w;
	}

	public double imagNorm() {
		return Math.sqrt(imagNormSqr());
	}

    public Quaternion negate() {
        return new Quaternion(-x, -y, -z, -w);
    }

    public Quaternion add(double r) {
        return new Quaternion(x + r, y, z, w);
    }

    public Quaternion add(Quaternion q) {
        return new Quaternion(x + q.x, y + q.y, z + q.z, w + q.w);
    }

    public Quaternion subtract(double r) {
        return new Quaternion(x - r, y, z, w);
    }

    public Quaternion subtract(Quaternion q) {
        return new Quaternion(x - q.x, y - q.y, z - q.z, w - q.w);
    }

    public Quaternion multiply(double r) {
        return new Quaternion(r * x, r * y, r * z, r * w);
    }

    public Quaternion multiply(Quaternion q) {
        return new Quaternion(x * q.x - y * q.y - z * q.z - w * q.w,
                x * q.y + y * q.x + z * q.w - w * q.z,
                x * q.z + z * q.x + w * q.y - y * q.w,
                x * q.w + w * q.x + y * q.z - z * q.y);
    }

    public Quaternion multiplySwitched(Quaternion q) {
        return q.multiply(this);    
    }

    public Quaternion sqr() {
        double hi = x * y;
        double hj = x * z;
        double hk = x * w;
        return new Quaternion(x * x - y * y - z * z - w * w, hi + hi, hj + hj, hk + hk);
    }

    public Quaternion cube() {
        double x2 = x * x;
        double y2 = y * y;
        double z2 = z * z;
        double w2 = w * w;
        double yy2 = y2 + z2 + w2;
        return new Quaternion(x * (x2 - 3.0 * yy2), (3.0 * x2 - yy2) * y, (3.0 * x2 - yy2) * z, (3.0 * x2 - yy2) * w);
    }

    public Quaternion inverse() {
        double den=normSqr();
        return new Quaternion(x / den, -y / den, -z / den, -w / den);
    }

    public Quaternion divide(double r) {
        return new Quaternion(x / r, y / r, z / r, w / r);
    }

    public Quaternion divide(Quaternion q) {
        return multiply(q.inverse());
    }

    public Quaternion divideSwitched(Quaternion q) {
        return q.inverse().multiply(this);
    }

    public Quaternion conjugate() {
        return new Quaternion(x, -y, -z, -w);
    }

    public Quaternion pow(int n) {
        if (n==2) return sqr();
        if (n==3) return cube();
        if (n==-1) return inverse();
        if (n==1) return this;
        if (n==0) return Quaternion.ONE;
        if (n<0) return inverse().pow(-n);
        if (n % 3 == 0) return cube().pow(n / 3);
        if (n % 2 == 0) return sqr().pow(n / 2);
        return multiply(sqr().pow((n-1) / 2));
    }

    public Quaternion getZero() {
        return ZERO;
    }

    public Quaternion getUnit() {
        return ONE;
    }

    public boolean isUnit() {
        return x == 1.0 && y == 0.0 && z == 0.0 && w == 0.0;
    }

    public String toString(NumberFormat nf, boolean parentheses, boolean negativeParentheses, boolean sign) {
        ComplexNumberFormatter cnf = new ComplexNumberFormatter(nf, parentheses, negativeParentheses, sign);
        cnf.addNumberString(x);
        cnf.addNumberString(y, "i");
        cnf.addNumberString(z, "j");
        cnf.addNumberString(w, "k");
        return cnf.toString();
    }

    public String toString(NumberFormat nf, boolean parentheses) {
        return toString(nf, parentheses, false, false);
    }

    public String toString(boolean parentheses, boolean negativeParentheses, boolean sign) {
        return toString(NumberFormatHelper.getDefaultNumberFormat(), parentheses, negativeParentheses, sign);
    }

    public String toString(boolean parentheses) {
        return toString(parentheses, false, false);
    }

    public String toString() {
        return toString(false);
    }

}

